core/src/test/scala/org/apache/spark/rdd/PipedRDDSuite.scala

```diff
@@ -83,6 +84,29 @@ class PipedRDDSuite extends SparkFunSuite with SharedSparkContext {
     }
   }
 
+  test("stdin writer thread should be exited when task is finished") {
+    assume(TestUtils.testCommandAvailable("cat"))
+    val nums = sc.makeRDD(Array(1, 2, 3, 4), 1).map { x =>
+      val obj = new Object()
+      obj.synchronized {
+        obj.wait() // make the thread waits here.
+      }
+      x
+    }
+
+    val piped = nums.pipe(Seq("cat"))
+
+    val result = piped.mapPartitions(_ => Array.emptyIntArray.iterator)
+
+    assert(result.collect().length === 0)
+
+    // collect stderr writer threads
+    val stderrWriterThreads = Thread.getAllStackTraces.keySet().asScala
+      .filter { _.getName.startsWith(PipedRDD.STDIN_WRITER_THREAD_PREFIX) }
```

Comment:
nit: `val stderrWriterThread = find { _.getname ...` and `assert(stderrWriterThread.isEmpty)`

Reply:
Will fix.